File: grasmere_routes/baseline_service.py
"""Baseline reconstruction orchestrator (in-process — no HTTP service).

  1. Pull live customers (status filter via customer_status_v).
  2. Drop those whose legacy_run_code is mail-order (~NR) or unparseable.
  3. Group survivors by (van colour × day of week) using decoded codes.
  4. Call the optimiser baseline reconstructor (nearest-neighbour per group).
  5. Persist baseline_snapshot + baseline_routes + baseline_route_stops.
  6. Mark this snapshot as is_current = TRUE; demote previous current.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import math
from typing import Any

from sqlalchemy import text

from .baseline import reconstruct_baseline
from .db import engine, query_df
from .queries import get_config, get_depot
from .run_code import RUN_DAYS, decode_run_code
from .schemas import (
    BaselineRequest,
    BaselineRouteSpec,
    CostParamsModel,
    LatLngModel,
    StopSpec,
)


def _cost_params_from_config(cfg: dict) -> CostParamsModel:
    return CostParamsModel(
        diesel_price_pence_per_litre=int(cfg["diesel_price_pence_per_litre"]),
        vehicle_mpg=float(cfg["default_mpg"]),
        driver_hourly_rate_pence=int(cfg["default_driver_hourly_rate_pence"]),
        avg_speed_kmh=float(cfg["avg_speed_kmh"]),
        service_min_per_stop=int(cfg["service_time_min_per_stop"]),
        depot_loading_min=int(cfg["depot_loading_time_min"]),
        vehicle_fixed_cost_per_day_pence=int(cfg["vehicle_fixed_cost_per_day_pence"]),
    )


def _config_hash(cfg: dict) -> str:
    payload = {k: (str(v) if not isinstance(v, (int, float, list)) else v) for k, v in cfg.items() if k != "updated_at"}
    return hashlib.sha256(json.dumps(payload, sort_keys=True, default=str).encode()).hexdigest()[:16]


def recompute_baseline() -> dict[str, Any]:
    cfg = get_config()
    depot = get_depot()

    live = query_df(
        """
        SELECT c.id::text AS id, c.legacy_run_code,
               c.delivery_lat, c.delivery_lng
        FROM customers c
        JOIN customer_status_v s ON s.customer_id = c.id
        WHERE c.active = TRUE AND s.status = 'live'
        """
    )

    excluded = 0
    groups: dict[tuple[str, int], list[dict]] = {}
    for _, r in live.iterrows():
        if r["delivery_lat"] is None or r["delivery_lng"] is None or (
            isinstance(r["delivery_lat"], float) and math.isnan(r["delivery_lat"])
        ):
            excluded += 1
            continue
        decoded = decode_run_code(r["legacy_run_code"])
        if decoded.is_mail_order or decoded.unparseable:
            excluded += 1
            continue
        for d in RUN_DAYS:
            colour = decoded.by_day.get(d)
            if not colour:
                continue
            groups.setdefault((colour, d), []).append(
                {
                    "id": r["id"],
                    "lat": float(r["delivery_lat"]),
                    "lng": float(r["delivery_lng"]),
                }
            )

    cost_params = _cost_params_from_config(cfg)
    req = BaselineRequest(
        depot=LatLngModel(lat=float(depot["lat"]), lng=float(depot["lng"])),
        baseline_routes=[
            BaselineRouteSpec(
                van_colour=colour,
                day_of_week=d,
                stops=[
                    StopSpec(
                        id=s["id"],
                        lat=s["lat"],
                        lng=s["lng"],
                        weight_kg=0,
                        crate_count=0,
                        service_minutes=int(cfg["service_time_min_per_stop"]),
                    )
                    for s in stops
                ],
            )
            for (colour, d), stops in groups.items()
        ],
        cost_params=cost_params,
    )
    resp = asyncio.run(reconstruct_baseline(req))

    # Persist
    cfg_hash = _config_hash(cfg)
    with engine().begin() as conn:
        conn.execute(
            text("UPDATE baseline_snapshot SET is_current = FALSE WHERE is_current = TRUE")
        )
        snap = conn.execute(
            text(
                """
                INSERT INTO baseline_snapshot (
                  config_hash, customer_count_included, customer_count_excluded,
                  total_distance_km, total_fuel_cost_pence, total_labour_cost_pence,
                  total_overhead_pence, total_cost_pence, total_stops,
                  weekly_cost_pence, annualised_cost_pence, is_current, notes
                ) VALUES (
                  :config_hash, :inc, :exc,
                  :dist, :fuel, :lab,
                  :ovh, :tot, :stops,
                  :wk, :ann, TRUE, :notes
                ) RETURNING id::text AS id
                """
            ),
            {
                "config_hash": cfg_hash,
                "inc": int(len(live) - excluded),
                "exc": int(excluded),
                "dist": float(resp.summary["total_distance_km"]),
                "fuel": int(sum(r.fuel_cost_pence for r in resp.baseline_routes)),
                "lab": int(sum(r.labour_cost_pence for r in resp.baseline_routes)),
                "ovh": int(sum(r.overhead_pence for r in resp.baseline_routes)),
                "tot": int(resp.summary["total_cost_pence"]),
                "stops": int(resp.summary["total_stops"]),
                "wk": int(resp.summary["weekly_cost_pence"]),
                "ann": int(resp.summary["annualised_cost_pence"]),
                "notes": (
                    f"Computed from {len(live)} live customers; {excluded} excluded "
                    "(mail-order, unparseable codes, or no geocode)"
                ),
            },
        ).first()
        snap_id = snap[0]

        for br in resp.baseline_routes:
            stops = max(len(br.stop_sequence), 1)
            route_id = conn.execute(
                text(
                    """
                    INSERT INTO baseline_routes (
                      snapshot_id, van_colour, day_of_week, stop_count,
                      distance_km, duration_min,
                      fuel_cost_pence, labour_cost_pence, overhead_pence,
                      total_cost_pence, cost_per_stop_pence
                    ) VALUES (
                      :sid, :van, :dow, :n, :km, :min,
                      :fuel, :lab, :ovh, :tot, :pps
                    ) RETURNING id::text AS id
                    """
                ),
                {
                    "sid": snap_id,
                    "van": br.van_colour,
                    "dow": br.day_of_week,
                    "n": stops,
                    "km": float(br.total_distance_km),
                    "min": int(br.total_duration_min),
                    "fuel": int(br.fuel_cost_pence),
                    "lab": int(br.labour_cost_pence),
                    "ovh": int(br.overhead_pence),
                    "tot": int(br.total_cost_pence),
                    "pps": int(round(br.total_cost_pence / stops)),
                },
            ).first()[0]

            for i, cid in enumerate(br.stop_sequence):
                leg_km = br.leg_distances_km[i] if i < len(br.leg_distances_km) else 0.0
                leg_min = br.leg_durations_min[i] if i < len(br.leg_durations_min) else 0
                fuel_share = int(round((leg_km / max(br.total_distance_km, 0.001)) * br.fuel_cost_pence))
                labour_share = int(round((leg_min / max(br.total_duration_min, 1)) * br.labour_cost_pence))
                overhead_share = br.overhead_pence // stops
                conn.execute(
                    text(
                        """
                        INSERT INTO baseline_route_stops (
                          baseline_route_id, customer_id, sequence,
                          km_to_stop, min_to_stop,
                          fuel_share_pence, labour_share_pence,
                          overhead_share_pence, direct_cost_pence
                        ) VALUES (
                          :brid, :cid, :seq, :km, :min,
                          :fuel, :lab, :ovh, :direct
                        )
                        """
                    ),
                    {
                        "brid": route_id,
                        "cid": cid,
                        "seq": i + 1,
                        "km": leg_km,
                        "min": leg_min,
                        "fuel": fuel_share,
                        "lab": labour_share,
                        "ovh": overhead_share,
                        "direct": fuel_share + labour_share + overhead_share,
                    },
                )

    return {
        "snapshot_id": snap_id,
        "customer_count_included": len(live) - excluded,
        "customer_count_excluded": excluded,
        "weekly_cost_pence": int(resp.summary["weekly_cost_pence"]),
        "annualised_cost_pence": int(resp.summary["annualised_cost_pence"]),
        "total_distance_km": float(resp.summary["total_distance_km"]),
    }

File: apps/web/lib/baseline-service.ts
/**
 * Server-side orchestration of the baseline reconstruction.
 *
 *   1. Pull live customers (status filter via customer_status_v).
 *   2. Drop those whose legacy_run_code is mail-order (~NR) or unparseable.
 *   3. Group survivors by (van colour × day of week) using decoded codes.
 *   4. Call the optimiser /baseline_cost endpoint.
 *   5. Persist baseline_snapshot + baseline_routes + baseline_route_stops.
 *   6. Mark this snapshot as is_current = TRUE; demote previous current.
 *
 * The "data hygiene" line item in /economics relies on this being correct.
 */
import { sql, eq, and, ne } from "drizzle-orm";
import { createHash } from "node:crypto";
import { db, schema } from "@/db/client";
import { execRows } from "./db-utils";
import { decodeRunCode, RUN_DAYS, type DayOfWeek, type VanColour } from "./run-code";
import { optimiser, type BaselineRequest } from "./optimiser-client";

export interface BaselineComputeResult {
  snapshotId: string;
  customerCountIncluded: number;
  customerCountExcluded: number;
  weeklyCostPence: number;
  annualisedCostPence: number;
  totalDistanceKm: number;
  routeBreakdown: { vanColour: string; dayOfWeek: number; totalCostPence: number }[];
}

interface LiveCustomerRow {
  id: string;
  legacy_run_code: string | null;
  delivery_lat: number | null;
  delivery_lng: number | null;
}

interface ConfigRow {
  diesel_price_pence_per_litre: number;
  default_mpg: string;
  default_driver_hourly_rate_pence: number;
  avg_speed_kmh: string;
  service_time_min_per_stop: number;
  depot_loading_time_min: number;
  vehicle_fixed_cost_per_day_pence: number;
}

export async function recomputeBaseline(): Promise<BaselineComputeResult> {
  // ---- read config ----
  const cfgRows = await execRows<ConfigRow>(sql`SELECT * FROM config WHERE id = 1`);
  const cfg = cfgRows[0];
  if (!cfg) throw new Error("config row missing — run db:migrate seed");

  const depotRow = await execRows<{ lat: number; lng: number }>(
    sql`SELECT lat, lng FROM depot ORDER BY name LIMIT 1`,
  );
  const depot = depotRow[0];
  if (!depot) throw new Error("no depot row — run db:migrate seed");

  // ---- pull live customers with geocodes ----
  const live = await execRows<LiveCustomerRow>(sql`
    SELECT c.id, c.legacy_run_code, c.delivery_lat, c.delivery_lng
    FROM customers c
    JOIN customer_status_v s ON s.customer_id = c.id
    WHERE c.active = TRUE AND s.status = 'live'
  `);

  // ---- group by van colour and day ----
  type Group = {
    vanColour: VanColour;
    dayOfWeek: DayOfWeek;
    customers: LiveCustomerRow[];
  };
  const groups = new Map<string, Group>();
  let excluded = 0;

  for (const c of live) {
    if (c.delivery_lat == null || c.delivery_lng == null) {
      excluded++;
      continue;
    }
    const decoded = decodeRunCode(c.legacy_run_code);
    if (decoded.isMailOrder || decoded.unparseable) {
      excluded++;
      continue;
    }
    for (const day of RUN_DAYS) {
      const colour = decoded.byDay[day];
      if (!colour) continue;
      const key = `${colour}|${day}`;
      const g =
        groups.get(key) ??
        ({ vanColour: colour, dayOfWeek: day, customers: [] } satisfies Group);
      g.customers.push(c);
      groups.set(key, g);
    }
  }

  // ---- call optimiser /baseline_cost ----
  const req: BaselineRequest = {
    depot: { lat: depot.lat, lng: depot.lng },
    baseline_routes: Array.from(groups.values()).map((g) => ({
      van_colour: g.vanColour,
      day_of_week: g.dayOfWeek,
      stops: g.customers.map((c) => ({
        id: c.id,
        lat: c.delivery_lat!,
        lng: c.delivery_lng!,
        weight_kg: 0,
        crate_count: 0,
        service_minutes: cfg.service_time_min_per_stop,
      })),
    })),
    cost_params: {
      diesel_price_pence_per_litre: cfg.diesel_price_pence_per_litre,
      vehicle_mpg: Number(cfg.default_mpg),
      driver_hourly_rate_pence: cfg.default_driver_hourly_rate_pence,
      avg_speed_kmh: Number(cfg.avg_speed_kmh),
      service_min_per_stop: cfg.service_time_min_per_stop,
      depot_loading_min: cfg.depot_loading_time_min,
      vehicle_fixed_cost_per_day_pence: cfg.vehicle_fixed_cost_per_day_pence,
    },
  };

  const resp = await optimiser.baselineCost(req);

  // ---- persist ----
  const cfgHash = hashConfig(cfg);

  // Demote previous current
  await db
    .update(schema.baselineSnapshot)
    .set({ isCurrent: false })
    .where(eq(schema.baselineSnapshot.isCurrent, true));

  const [snap] = await db
    .insert(schema.baselineSnapshot)
    .values({
      configHash: cfgHash,
      customerCountIncluded: live.length - excluded,
      customerCountExcluded: excluded,
      totalDistanceKm: resp.summary.total_distance_km.toString(),
      totalFuelCostPence: resp.baseline_routes.reduce((a, r) => a + r.fuel_cost_pence, 0),
      totalLabourCostPence: resp.baseline_routes.reduce((a, r) => a + r.labour_cost_pence, 0),
      totalOverheadPence: resp.baseline_routes.reduce((a, r) => a + r.overhead_pence, 0),
      totalCostPence: resp.summary.total_cost_pence,
      totalStops: resp.summary.total_stops,
      weeklyCostPence: resp.summary.weekly_cost_pence,
      annualisedCostPence: resp.summary.annualised_cost_pence,
      isCurrent: true,
      notes: `Computed from ${live.length} live customers; ${excluded} excluded (mail-order, unparseable codes, or no geocode)`,
    })
    .returning({ id: schema.baselineSnapshot.id });

  for (const r of resp.baseline_routes) {
    const [insertedRoute] = await db
      .insert(schema.baselineRoutes)
      .values({
        snapshotId: snap!.id,
        vanColour: r.van_colour,
        dayOfWeek: r.day_of_week,
        stopCount: r.stop_sequence.length,
        distanceKm: r.total_distance_km.toString(),
        durationMin: r.total_duration_min,
        fuelCostPence: r.fuel_cost_pence,
        labourCostPence: r.labour_cost_pence,
        overheadPence: r.overhead_pence,
        totalCostPence: r.total_cost_pence,
        costPerStopPence: Math.round(r.total_cost_pence / Math.max(r.stop_sequence.length, 1)),
      })
      .returning({ id: schema.baselineRoutes.id });

    // Per-stop allocation rows
    for (let i = 0; i < r.stop_sequence.length; i++) {
      const cid = r.stop_sequence[i]!;
      const legKm = r.leg_distances_km[i] ?? 0;
      const legMin = r.leg_durations_min[i] ?? 0;
      const fuelShare = Math.round((legKm / Math.max(r.total_distance_km, 0.001)) * r.fuel_cost_pence);
      const labourShare = Math.round((legMin / Math.max(r.total_duration_min, 1)) * r.labour_cost_pence);
      const overheadShare = Math.floor(r.overhead_pence / Math.max(r.stop_sequence.length, 1));
      await db.insert(schema.baselineRouteStops).values({
        baselineRouteId: insertedRoute!.id,
        customerId: cid,
        sequence: i + 1,
        kmToStop: legKm.toString(),
        minToStop: legMin,
        fuelSharePence: fuelShare,
        labourSharePence: labourShare,
        overheadSharePence: overheadShare,
        directCostPence: fuelShare + labourShare + overheadShare,
      });
    }
  }

  return {
    snapshotId: snap!.id,
    customerCountIncluded: live.length - excluded,
    customerCountExcluded: excluded,
    weeklyCostPence: resp.summary.weekly_cost_pence,
    annualisedCostPence: resp.summary.annualised_cost_pence,
    totalDistanceKm: resp.summary.total_distance_km,
    routeBreakdown: resp.baseline_routes.map((r) => ({
      vanColour: r.van_colour,
      dayOfWeek: r.day_of_week,
      totalCostPence: r.total_cost_pence,
    })),
  };
}

function hashConfig(cfg: ConfigRow): string {
  return createHash("sha256")
    .update(JSON.stringify(cfg))
    .digest("hex")
    .slice(0, 16);
}

export async function getCurrentBaseline() {
  const rows = await db
    .select()
    .from(schema.baselineSnapshot)
    .where(eq(schema.baselineSnapshot.isCurrent, true))
    .limit(1);
  return rows[0] ?? null;
}

export async function getCurrentBaselineRoutes(snapshotId: string) {
  return db
    .select()
    .from(schema.baselineRoutes)
    .where(eq(schema.baselineRoutes.snapshotId, snapshotId))
    .orderBy(schema.baselineRoutes.dayOfWeek, schema.baselineRoutes.vanColour);
}

// silence unused-import lint (and, ne) — kept for future filters
void and; void ne;
